mdal/mdal_memory_data_model.hpp

```diff
@@ -171,22 +171,36 @@ namespace MDAL
   class MemoryMesh: public Mesh
   {
     public:
+      //! Construct an empty mesh
       MemoryMesh( const std::string &driverName,
-                  size_t verticesCount,
-                  size_t edgesCount,
-                  size_t facesCount,
                   size_t faceVerticesMaximumCount,
-                  BBox extent,
                   const std::string &uri );
+
       ~MemoryMesh() override;
 
       std::unique_ptr<MDAL::MeshVertexIterator> readVertices() override;
       std::unique_ptr<MDAL::MeshEdgeIterator> readEdges() override;
       std::unique_ptr<MDAL::MeshFaceIterator> readFaces() override;
 
-      Vertices vertices;
-      Faces faces;
-      Edges edges;
+      const Vertices &vertices() const {return mVertices;}
+      const Faces &faces() const {return mFaces;}
+      const Edges &edges() const {return mEdges;}
+
+      void setVertices( Vertices vertices );
```

Comment:
it is intentionally passed by value to avoid copying rvalue : https://clang.llvm.org/extra/clang-tidy/checks/modernize-pass-by-value.html
More efficient in this context.

Reply:
is is working across all supported compilers (gcc, clang, visual studio, etc?) I would rather stick with the const reference, which is what you usually expect when reading setter code?

Reply:
It a very documented way to avoid the costly copy with rvalue. In this context, I think it is very interesting to avoid copy of very large std::vector.
The other way is to implement two methods:
- one with const reference for lvalue
- one with universal reference that can move the std::vector
`setVertices(std::vector<Vertex> &&vertices);`
IMHO, one method is enough and easier to maintain.

Reply:
@wonder-sk

Reply:
I am still a bit lost in the whole move constructor semantics - if Vincent is confident this is more efficient - let's stick to that...

Reply:
ok, lets do std::move, but would it be possible to encapsulate it in the setter directly as suggested here https://stackoverflow.com/a/21227582 in point 2)?

```
void setVertices(std::vector<Vertex> vertices) {
  mVertices = std::move(vertices);
}
```

and the caller would be just 

```
 myMesh->setVertices(myVertices);
```

With this you also let compiler decide if you CAN move the array ...
Also add to docs of the setter it uses move semantics.

Thanks

Reply:
the std::move is already encapsulate in the setter, and the compiler choose depending rvalue/lvalue : 
- If the argument is a rvalue, no copy, only move, but no problem because rvalue will not be used on the caller side.
- If the argmument is a lvalue, one copy to pass the argument, and one move in the setter. And no problem to use again the argument on the caller side because a copy has been done.

On the caller side, here, WE use std::move to transform the lvalue to rvalue, so avoiding a copy. And that after being SURE the lvalue is not used after this calling. 

OK for adding a doc.

Reply:
i think that if we skip std::move in caller and move it inside the setter, the resulted optimized code will be the same, but we not not need std::move all around the code base and also in the future if we want to revert to const vector& it will just work?

Reply:
just if you can remove std::move call from the callers and see if there is still the same performance benefit?

Reply:
Little benchmark,
First the code:

```
struct Vertices
{
  double x, y, z;
};


class VectorContainer
{
  public:

    void setVerticesWithMove( std::vector<Vertices> vertices )
    {
      mVertices = std::move( vertices );
    }

    void setVerticesWithoutMove( const std::vector<Vertices> vertices )
    {
      mVertices = vertices;
    }

    void clear()
    {
      mVertices.clear();
    }

  private:
    std::vector<Vertices> mVertices;
};

int main( int argc, char *argv[] )
{
  QCoreApplication a( argc, argv );



  std::vector<Vertices> vertices;

  //populateVertices
  std::cout << "populate " << std::endl;
  size_t count = 10000000;
  vertices.resize( count );
  for ( size_t i = 0; i < count; ++i )
  {
    vertices[i] = Vertices( {1.2, 2.3, 3.4} );
  }

  QTime time;
  time.restart();

  VectorContainer vectorContainer;

  // by const ref
  for ( int i = 0; i < 10; i++ )
  {
    vectorContainer.clear();
    vectorContainer.setVerticesWithoutMove( vertices );
  }
  std::cout << "Copy without move: " << time.elapsed() << " ms" << std::endl;
  time.restart();
  // by const ref
  for ( int i = 0; i < 10; i++ )
  {
    vectorContainer.clear();
    vectorContainer.setVerticesWithMove( vertices );
  }
  std::cout << "Copy with move in setter: " << time.elapsed() << " ms" << std::endl;

  time.restart();
  // by const ref
  for ( int i = 0; i < 10; i++ )
  {
    vectorContainer.clear();
    vectorContainer.setVerticesWithMove( std::move( vertices ) );
  }
  std::cout << "Copy with move in setter and caller: " << time.elapsed() << " ms" << std::endl;

  return a.exec();

}
```

The resutl is :
![image](https://user-images.githubusercontent.com/7416892/83038750-50dd8280-a00b-11ea-9dcc-17a44f4ad303.png)


Big winner is : std::move in the setter AND in the caller.

Reply:
looks good, but would it be possible to create different vectorContainer for each setter? 

since in this test the vectorContainer.setVerticesWithMove( vertices ); is not used correctly. The compiler see that you call vectorContainer.clear() afterwards, so it knows it CANNOT use std:move (since the container is modified), so it transforms to plain deep copy. 

So in this test only copy without mode and copy with move in setter and caller are correctly timed.

Reply:
Corrected and more complete code:
```
struct Vertices
{
  double x, y, z;
};


class VectorContainer
{
  public:

    void setVerticesWithConstef( const std::vector<Vertices> &vertices )
    {
      mVertices = vertices;
    }

    void setVerticesWithMoveConstRef( const std::vector<Vertices> &vertices )
    {
      mVertices = std::move( vertices );
    }


    void setVerticesWithMove( std::vector<Vertices> vertices )
    {
      mVertices = std::move( vertices );
    }

    void clear()
    {
      mVertices.clear();
    }

  private:
    std::vector<Vertices> mVertices;
};

int main( int argc, char *argv[] )
{
  QCoreApplication a( argc, argv );



  std::vector<Vertices> vertices;

  //populateVertices
  std::cout << "populate " << std::endl;
  size_t count = 10000000;
  vertices.resize( count );
  for ( size_t i = 0; i < count; ++i )
  {
    vertices[i] = Vertices( {1.2, 2.3, 3.4} );
  }

  QTime time;
  time.restart();

  VectorContainer vectorContainer1;
  VectorContainer vectorContainer2;
  VectorContainer vectorContainer3;
  VectorContainer vectorContainer4;

  for ( int i = 0; i < 10; i++ )
  {
    vectorContainer1.clear();
    vectorContainer1.setVerticesWithConstef( vertices );
  }
  std::cout << "Set with const ref: " << time.elapsed() << " ms" << std::endl;

  time.restart();
  for ( int i = 0; i < 10; i++ )
  {
    vectorContainer2.clear();
    vectorContainer2.setVerticesWithMoveConstRef( vertices );
  }
  std::cout << "Set with const ref and move in setter: " << time.elapsed() << " ms" << std::endl;

  time.restart();
  for ( int i = 0; i < 10; i++ )
  {
    vectorContainer3.clear();
    vectorContainer3.setVerticesWithMove( vertices );
  }
  std::cout << "Set with value and move in setter: " << time.elapsed() << " ms" << std::endl;

  time.restart();
  for ( int i = 0; i < 10; i++ )
  {
    vectorContainer4.clear();
    vectorContainer4.setVerticesWithMove( std::move( vertices ) );
  }
  std::cout << "Set with value,  move in setter and caller: " << time.elapsed() << " ms" << std::endl;

  return a.exec();
}
```

Result:
![image](https://user-images.githubusercontent.com/7416892/83046435-052fd680-a015-11ea-8554-e0db0492d6e1.png)
It seems the solution with move only in setter is not good. Maybe there are two copies. 
But with move in setter and caller, still the big winner. As the setter for faces, edges and vertices are used only for massive transfer of data that are not used after in the caller, I still prefer this solution.

Reply:
this is getting interesting...
if you do not mind , can you also try this last one?
```
std::vector<Vertices> vertices1;
std::vector<Vertices> vertices2;
std::vector<Vertices> vertices3;
std::vector<Vertices> vertices4;
```

Reply:
with 
```
std::vector<Vertices> vertices1;
std::vector<Vertices> vertices2;
std::vector<Vertices> vertices3;
std::vector<Vertices> vertices4;
```
![image](https://user-images.githubusercontent.com/7416892/83071305-e7756800-a03a-11ea-80da-64701cc2c71a.png)
But I don't understand why the third is so high...

Reply:
Ok, my bench was wrong because with the for loop to set the Vertices in the verticesContainer and the std::move, the vector was moved in the first pass, so not present for other pass...
So new code without for loop but more vertices :+1: 
```

struct Vertices
{
  double x, y, z;
};


class VectorContainer
{
  public:

    void setVerticesWithConstef( const std::vector<Vertices> &vertices )
    {
      mVertices = vertices;
    }

    void setVerticesWithMoveConstRef( const std::vector<Vertices> &vertices )
    {
      mVertices = std::move( vertices );
    }


    void setVerticesWithMove( std::vector<Vertices> vertices )
    {
      mVertices = std::move( vertices );
    }

    void clear()
    {
      mVertices.clear();
    }

    size_t count()
    {
      return mVertices.size();
    }

  private:
    std::vector<Vertices> mVertices;
};

int main( int argc, char *argv[] )
{
  QCoreApplication a( argc, argv );



  std::vector<Vertices> vertices1;
  std::vector<Vertices> vertices2;
  std::vector<Vertices> vertices3;
  std::vector<Vertices> vertices4;

  //populateVertices
  std::cout << "populate " << std::endl;
  size_t count = 50000000;
  vertices1.resize( count );
  vertices2.resize( count );
  vertices3.resize( count );
  vertices4.resize( count );
  for ( size_t i = 0; i < count; ++i )
  {
    vertices1[i] = Vertices( {1.2, 2.3, 3.4} );
    vertices2[i] = Vertices( {1.2, 2.3, 3.4} );
    vertices3[i] = Vertices( {1.2, 2.3, 3.4} );
    vertices4[i] = Vertices( {1.2, 2.3, 3.4} );
  }

  QTime time;


  VectorContainer vectorContainer1;
  VectorContainer vectorContainer2;
  VectorContainer vectorContainer3;
  VectorContainer vectorContainer4;

  time.restart();
  vectorContainer1.setVerticesWithConstef( vertices1 );

  std::cout << "Set with const ref: " << time.elapsed() << " ms" << std::endl;
  std::cout << "Vertices count: " << vectorContainer1.count() << std::endl;

  time.restart();
  vectorContainer2.setVerticesWithMoveConstRef( vertices2 );

  std::cout << "Set with const ref and move in setter: " << time.elapsed() << " ms" << std::endl;
  std::cout << "Vertices count: " << vectorContainer2.count() << std::endl;

  time.restart();
  vectorContainer3.setVerticesWithMove( vertices3 );

  std::cout << "Set with value and move in setter: " << time.elapsed() << " ms" << std::endl;
  std::cout << "Vertices count: " << vectorContainer3.count() << std::endl;

  time.restart();
  vectorContainer4.setVerticesWithMove( std::move( vertices4 ) );

  std::cout << "Set with value,  move in setter and caller: " << time.elapsed() << " ms" << std::endl;
  std::cout << "Vertices count: " << vectorContainer4.count() << std::endl;

  return a.exec();
}
```
And results:
![image](https://user-images.githubusercontent.com/7416892/83072468-cf9ee380-a03c-11ea-929c-e96584632039.png)

Reply:
Last episode... The third one was so hight because ... it used my swap... Not enought RAM when it uses this third one.
So, I did a last benchmark using many path and for only three type of setter :
- argument by const ref
- by value with std::move in the setter
- by value with std::move in the setter and in the caller.

With 50 000 000 vertices.
And the result are as expected (average value with 100 calls) :
![image](https://user-images.githubusercontent.com/7416892/83084893-4e554a00-a058-11ea-9a62-c3a294eb87cc.png)

So passing with value with a std::move in the setter has the following advantages:
- if the argument is a rvalue, this call is no time consuming (or very few)
- if the argument is a lvalue, this call is time consuming like a setter with const ref (only one copy with std::move in the setter)
- as this is passed by value, the lvalue is copied in the setter and can't be modified in the setter, the rvalue is changed, but not important because not used after calling the setter in any case
- the user has the choice to pass the argument with lvalue or rvalue (with std::move that consist to simply converts lvalue to rvalue, or any other functions that produce the std::vector as a rvalue)

This method has the following disavantage:
- no like the usual signature setters

Reply:
ok, thanks for clarifying the things.